### PR TITLE
vconsole-setup: handle the case where the vc is in KD_GRAPHICS mode m…

### PR DESCRIPTION
**Pull request by fbuihuu**
…ore gracefully

Regardless of whether a vc path is passed, the behavior of systemd-vconsole-setup wasn't ideal when either the passed vc or /dev/tty1 was in graphics mode.

When a vc in graphics mode was passed, no message was emitted despite the fact that the font settings couldn't be applied. The previous code might have assumed that setfont(8) would throw a warning but that's not case.

When no argument was passed, systemd-vconsole-setup was supposed to automatically select a valid tty, init it and copy the font setting to the remaining ttys. However if the selected virtual console was in KD_GRAPHICS mode the initialization of the font failed not only for the selected source vc but for all of them.

### CODE REVIEW

File: src/vconsole/vconsole-setup.c [hunk 1]
```diff
@@ -470,24 +482,14 @@ static void setup_remaining_vcs(int src_fd, unsigned src_idx, bool utf8) {
                 if (cfo.op != KD_FONT_OP_SET)
                         continue;
 
+                if (verify_vc_kdmode(fd_d) < 0) {
+                        log_debug("KD_FONT_OP_SET skipped: tty%u is not in text mode", i);
```
Comment (poettering) [inline src/vconsole/vconsole-setup.c]: please use log_debug_errno() and pass error code as first argument, even if you don't output it via %m. our log functions generated structured logs if possible and will include the error numer in a structured log field if provided. hence: *always* pass the error code to the log functions if you have one, regardless of using %m or not

File: src/vconsole/vconsole-setup.c [hunk 2]
```diff
@@ -572,6 +582,13 @@ static int verify_source_vc(char **ret_path, const char *src_vc) {
         if (r < 0)
                 return log_error_errno(r, "Virtual console %s is not in K_XLATE or K_UNICODE: %m", src_vc);
 
+        /* setfont(8) silently ignores when the font can't be applied due to the vc being in
+         * KD_GRAPHICS. Hence we continue to accept this case however we now let the user know that the vc
+         * will be initialized only partially.*/
+        r = verify_vc_kdmode(fd);
+        if (r < 0)
+                log_notice("Virtual console %s is not in KD_TEXT, font settings won't be applied.", src_vc);
```
Comment (poettering) [inline src/vconsole/vconsole-setup.c]: as above

Comment (poettering) [reply src/vconsole/vconsole-setup.c]: and i'd tone down the message a bit. What you describe here is how setfont() operates today, maybe this is changed one day. i.e. just say "font settings likely won't be applied".

File: src/vconsole/vconsole-setup.c [hunk 3]
```diff
@@ -231,6 +231,18 @@ static int verify_vc_kbmode(int fd) {
         return IN_SET(curr_mode, K_XLATE, K_UNICODE) ? 0 : -EBUSY;
 }
 
+static int verify_vc_kdmode(int fd) {
```
Comment (poettering) [inline src/vconsole/vconsole-setup.c]: "kdmode" is not a term the console layer uses in its ioctl iface, is it? maybe don't introduce such cryptic new terms. just call this "verify_vc_display_mode()"?

**Comment by poettering** [comment]
looks ok, just some nitpicks

**Comment by fbuihuu** [comment]
Thanks for the review. New version force pushed.

File: src/vconsole/vconsole-setup.c [hunk 4]
```diff
@@ -231,6 +231,18 @@ static int verify_vc_kbmode(int fd) {
         return IN_SET(curr_mode, K_XLATE, K_UNICODE) ? 0 : -EBUSY;
 }
 
+static int verify_vc_display_mode(int fd) {
+        int mode;
+
+        /* Similarly the vc is likely busy if it is in KD_GRAPHICS mode. If it's not the case and it's been
+         * left in graphics mode, the kernel will refuse to operate on the font settings anyway. */
+
```
Comment (YHNdnzj) [inline src/vconsole/vconsole-setup.c:243]: assert(fd >= 0)

File: src/vconsole/vconsole-setup.c [hunk 5]
```diff
@@ -535,13 +538,21 @@ static int find_source_vc(char **ret_path, unsigned *ret_idx) {
...
                 if (r < 0) {
                         log_debug_errno(r, "Failed to check VC %s keyboard mode: %m", path);
                         RET_GATHER(err, r);
                         continue;
                 }
 
+                r = verify_vc_display_mode(fd);
+                if (r < 0) {
+                        log_debug_errno(r, "Failed to check VC %s display mode: %m", path);
+                        RET_GATHER(err, r);
```
Comment (YHNdnzj) [inline src/vconsole/vconsole-setup.c]: Please merge this with the log_debug_errno above

Comment (fbuihuu) [reply src/vconsole/vconsole-setup.c]: I'm not sure to understand. The previous check is for the keyboard mode whereas this one is for the display mode. How can these be merged ?

Comment (YHNdnzj) [reply src/vconsole/vconsole-setup.c]: RET_GATHER(err, log_debug_errno(r, ...))

Comment (fbuihuu) [reply src/vconsole/vconsole-setup.c]: ah ok. Since none of the rest of the code do that currently, can't this be done later in a separate PR ?

Comment (fbuihuu) [reply src/vconsole/vconsole-setup.c]: This should be addressed by the second patch I just pushed.

File: src/vconsole/vconsole-setup.c [hunk 6]
```diff
@@ -470,24 +486,15 @@ static void setup_remaining_vcs(int src_fd, unsigned src_idx, bool utf8) {
...
-                        if (ioctl(fd_d, KDGETMODE, &mode) >= 0 && mode != KD_TEXT)
-                                log_debug("KD_FONT_OP_SET skipped: tty%u is not in text mode", i);
-                        else
-                                log_warning_errno(last_errno, "KD_FONT_OP_SET failed, fonts will not be copied to tty%u: %m", i);
+                        log_debug_errno(r, "KD_FONT_OP_SET skipped: tty%u is not in text mode", i);
+                        continue;
+                }
 
+                r = ioctl(fd_d, KDFONTOP, &cfo);
+                if (r < 0) {
```
Comment (poettering) [inline src/vconsole/vconsole-setup.c]: if you touch this you could actually shorten this to:

```c
if (ioctl(…) < 0) {
```

using `r` here is a bit pointless, it carries no useful value for us.

**Comment by poettering** [comment]
i figure @YHNdnzj pushed another patch into this, but that got overriden now?

**Comment by YHNdnzj** [comment]
> i figure @YHNdnzj pushed another patch into this, but that got overriden now?

Hmm, I didn't?

**Comment by poettering** [comment]
Oops, sorry, got confused. I read this:

https://github.com/systemd/systemd/pull/30491#discussion_r1430031007

And i thought it was you, but now i realize that @fbuihuu just posted twice in a row. sorry for the confusion

File: src/vconsole/vconsole-setup.c (hunk 6 above)
Comment (fbuihuu) [reply src/vconsole/vconsole-setup.c]: Fixed.
